keyvi/include/keyvi/dictionary/fuzzy_match.h

```diff
@@ -0,0 +1,131 @@
+/* * keyvi - A key value store.
+ *
+ * Copyright 2018   Narek Gharibyan<narekgharibyan@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * fuzzy_match.h
+ *
+ *  Created on: February 11, 2018
+ *      Author: Narek Gharibyan <narekgharibyan@gmail.com>
+ */
+
+#ifndef KEYVI_DICTIONARY_FUZZY_MATCH_H_
+#define KEYVI_DICTIONARY_FUZZY_MATCH_H_
+
+#include <string>
+#include <vector>
+
+#include "stringdistance/levenshtein.h"
+
+#include "dictionary/fsa/automata.h"
+#include "dictionary/fsa/codepoint_state_traverser.h"
+#include "dictionary/fsa/traverser_types.h"
+#include "dictionary/match.h"
+#include "dictionary/util/utf8_utils.h"
+
+namespace keyvi {
+namespace dictionary {
+
+class FuzzyMatch {
+ public:
+  FuzzyMatch(fsa::automata_t fsa, const std::string& query, size_t max_edit_distance)
+      : fsa_(std::move(fsa)), query_(query), max_edit_distance_(max_edit_distance) {
+    std::vector<int> codepoints;
+    utf8::unchecked::utf8to32(query.begin(), query.end(), back_inserter(codepoints));
+
+    query_char_length_ = codepoints.size();
+    metric_ptr_.reset(new stringdistance::Levenshtein(codepoints, 20, max_edit_distance_));
```

Comment:
I suggest to move that down after the prefix has been matched, so in case prefix does not match we avoid initialization cost